Add AutoBoxingCount and MethodInvocationCount measurers

kjanosz opened the pull request:
Add invocation counting measurers.
Add `prepareContext` and `cleanupContext` Measurer methods.
Modify executors to call above methods.

kjanosz commented:
@axel22: After we iterate over it, two things, which could be done, come to my mind are:

- add additional `ClassMatcher` that allows to instrument all methods which are implemented in subclasses of a given class (method from base class/trait is `abstract`)
- repackage `ASM`

What do you think about it?

kjanosz commented:
Also, I've noticed some strange behaviour, which looks like a bug for me:
When setting `Key.verbose` to `false` running any tests that involves spawning other JVM results in an exception (then the default `jvmflags` is `-Xmx2048m ...` that is set in `Context.topLevel`).
I guess that it could be easily solved by changing `jvmflags` key type from `String` to `List[String]`.
What do you think about it?

axel22 commented on scalameter-core/src/main/scala/org/scalameter/InvocationCounting.scala:
```diff
@@ -0,0 +1,64 @@
+      sys.error("Measurer.prepareContext must be called before Measurer.measure")).length
+
+    while (iteration < measurements) {
+      value = regen()
```
Similarly, if the run is deterministic, you can compute the value only once at the beginning.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/InvocationCounting.scala:
```diff
@@ -0,0 +1,64 @@
+    val numMethods = context.goe(exec.measurers.methodInvocationLookupTable,
+      sys.error("Measurer.prepareContext must be called before Measurer.measure")).length
+
+    while (iteration < measurements) {
```
This measurer is free to ignore the `measurements` argument, since we expect that it will be invoked for deterministic runs, where the invocation counts are constant.

Perhaps you can introduce a flag `Key.exec.assumeDeterministicRun` or something like that, which defaults to `true`, and assumes that `measurements` can be ignored.

kjanosz replied:
Do you mean that it should run only once? (So no while loop) Or the arbitrary number of times? (assuming that the `Key.exec.assumeDeterministicRun` is set to `true`)?

axel22 replied:
Yes - if the snippet is expected to be deterministic (`assumeDeterministicRun -> true`), no need for a loop, so `measurements` can be ignored.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/InvocationCounting.scala:
```diff
@@ -0,0 +1,64 @@
+import scala.collection.Seq
+import org.scalameter.execution.invocation.InvocationCountMatcher
+import org.scalameter.execution.invocation.instrumentation.{Instrumentation, MethodInvocationCounter}
+import org.scalameter.Key._
```
Nit: capital `K` comes earlier in ASCII alphabetic order.

axel22 commented:
The additional `ClassMatcher` sounds good!
What do you mean by repackaging `ASM`?
About the `Key.verbose -> false`, what is the exception and the stack trace that occurs?

axel22 commented on scalameter-core/src/main/scala/org/scalameter/InvocationCounting.scala:
```diff
@@ -0,0 +1,64 @@
+    invocations
+  }
+
+  /** Creates a jar with instrumented classes, prepends it to a classpath key and saves the internal method lookup table. */
```
"..., adds it to the context under the key `Key.classpath`, and ..."

axel22 commented on scalameter-core/src/main/scala/org/scalameter/Measurer.scala:
```diff
@@ -2,9 +2,14 @@ package org.scalameter
+import scala.compat._
+import scala.runtime.BoxesRunTime
+import scala.util.matching.Regex
+import org.scalameter.execution.invocation.InvocationCountMatcher
```
`o` comes before `s`.

Btw, thanks for expanding these imports!

kjanosz replied:
I was always thinking that I should place internal imports as the last :-)
So probably, they are unfortunately also in such order in other places when I was touching the code (but I'm not sure about it and don't know how to grep it easily)

axel22 replied:
Yes, this is how it used to be organized in the past, but I don't think that was a good decision any more.
Code is often migrated between projects, and what used to be internal can become external at one point. And when names change, you have to refactor the imports anyway, at which point you hopefully reorder them as well.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/InvocationCounting.scala:
```diff
@@ -0,0 +1,64 @@
+  /** Creates a jar with instrumented classes, prepends it to a classpath key and saves the internal method lookup table. */
+  override def prepareContext(context: Context): Context = {
+    val jar = File.createTempFile(s"scalameter-bench-", "-instrumented.jar")
+    jar.deleteOnExit()
```
Since `prepareContext` is called at the beginning of the run, before the separate VM is started, this file is not deleted until you exit the host VM. Meaning that if you call the test several times from SBT, these will leak.
You should probably delete the file manually (in addition to this call) at the end of the `run` method in the `Executor`.

kjanosz replied:
Right, I guess that it means that I should probably create `private[scalameter] instrumentedJarPath` key that will contain location of this file?

axel22 replied:
You'll probably have to. But you'll also have to introduce a method `afterExecution`, which does `Measurer`-specific cleanups.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/InvocationCounting.scala:
```diff
@@ -0,0 +1,64 @@
+
+    val lookupTable = Instrumentation.writeInstrumentedClasses(ctx = context, matcher = matcher, to = jar)
+    val classPath = context.goe(classpath, utils.ClassPath.default)
+    Context(exec.measurers.methodInvocationLookupTable -> lookupTable, classpath -> (jar +: classPath))
```
My understanding is that this should work correctly - earlier entries in the classpath override the entries later in the classpath: http://docs.oracle.com/javase/8/docs/technotes/tools/findingclasses.html#jarclass

kjanosz replied:
Yes, exactly.

axel22 replied:
Although, another strategy comes to mind. Could you:
1. create a `URLClassLoader` with the newly instrumented JAR file.
2. instruct the classloader to create the snippet class.
3. 

There are 2 advantages:
- the instrumented classes in the jar take precedence over the classes in the parent classloader
- the `LocalExecutor` then works correctly - no need to spawn a new VM

kjanosz replied:
I don't know about any other methods for setting `ClassLoder` in runtime then `setContextClassLoader` from the `Thread` and I don't know if this would work in that case. 

So to be sure that it won't block other parts maybe I could check it later, when we iterate over this PR?

axel22 replied:
I think you want to use the `ClassLoader` to load a class, and then instantiate it and run some of its methods to ensure that a piece of logic runs with all the classes from that classloader.

But, yes, we can iterate on that later if there's time. It would require some thinking and figuring things out.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/InvocationCounting.scala:
```diff
@@ -0,0 +1,64 @@
+    Context(exec.measurers.methodInvocationLookupTable -> lookupTable, classpath -> (jar +: classPath))
+  }
+
+  /** Restores the context before the concatenation with the context from [[prepareContext)]].
```
Please remove the typo with the closing paren in `prepareContext`.

kjanosz commented:
@axel22:
I've created an issue for this behaviour with `Key.verbose` set to `false` - please see #136.

As for the repackaging ASM, I've meant to change the base package of it and embed it in ScalaMeter to avoid API conflicts with user that might use ASM (I've mentioned about it in the proposal doc). 
However, I've looked today at the possible solutions and it's not so quick task to do, so given the time limit and fact that other improvements might be more desirable, I'm now not sure about it.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/Measurer.scala:
```diff
@@ -12,6 +17,12 @@ trait Measurer extends Serializable {
   def measure[T, U](context: Context, measurements: Int, setup: T => Any, tear: T => Any, regen: () => T, snippet: T => Any): Seq[Double]
   def units: String
+
+  /** Prepares fresh context that should be concatenated with the setup context. */
```
Could you please improve this comment:
- which context is this method called on?
- why is the method useful?
- how does the method change the context?
- when is this method called?
- who calls the method?

axel22 commented on scalameter-core/src/main/scala/org/scalameter/Measurer.scala:
```diff
@@ -12,6 +17,12 @@ trait Measurer extends Serializable {
+  /** Prepares fresh context that should be concatenated with the setup context. */
+  def prepareContext(context: Context): Context = Context()
+
+  /** Cleans context from additional [[org.scalameter.Context]] returned by [[prepareContext)]]. */
```
Could you similarly improve this comment?

axel22 commented on scalameter-core/src/main/scala/org/scalameter/utils/ClassPath.scala:
```diff
@@ -20,6 +20,12 @@ class ClassPath private (val paths: Seq[File]) extends Serializable {
     new ClassPath(location +: paths)
   }
 
+  /** Returns classpath without first location.
```
What if the classpath is empty?

kjanosz replied:
Then it throws an exception (should I update comment?). I know, not very good, but this method should be used only for cleaning context after adding instrumented jar. 
So maybe limiting access scope could be a good idea?

axel22 replied:
I would just update the comment. I would also add a method that checks if the classpath is empty (i.e. has no paths). Otherwise, using this function always requires `try-catch` blocks.

kjanosz replied:
I've removed this method (`ClassPath.tail`) since it wasn't used after all.

axel22 commented on src/test/scala/org/scalameter/MeasurerTest.scala:
```diff
@@ -0,0 +1,20 @@
+import org.scalameter.execution.JvmRunner
+
+
+abstract class MeasurerTest[M <: Measurer] extends FunSuite with Matchers {
```
Could you additionally add a few integration-like tests in the spirit of `JavaPerformanceTestTest.scala`?

kjanosz replied:
I've added `BoxingCountTest.scala` and `InvocatonCountTest.scala`

axel22 commented on src/main/scala/org/scalameter/Executor.scala:
```diff
@@ -1,18 +1,21 @@
 
 trait Executor {
 
+  def measurer: Measurer
```
To be consistent, you can add a `warmer` here as well.

kjanosz replied:
What about `aggregator`? Should I also add it?

axel22 replied:
This is something I think we should ultimately get rid of, so I would say no.

axel22 commented on src/main/scala/org/scalameter/Executor.scala:
```diff
@@ -1,18 +1,21 @@
+  def measurer: Measurer
+
   def run[T](setuptree: Tree[Setup[T]], reporter: Reporter, persistor: Persistor): Tree[CurveData] = {
+    val context = measurer.prepareContext(setuptree.context)
```
It seems to me that the initial `prepareContext` call must:
a) do some side-effecting actions to prepare a testing environment
b) return a couple of values that describe the prepared testing environment

# Alternative A

Initially, I was hoping we could replace the top-level context altogether before it gets used to create `Tree[Setup[T]]`, and that `BasePerformanceTest#defaultConfig` could be used for this purpose. Now I see that this is broken - `defaultConfig` is just ignored.

The way to fix it is to use `defaultConfig` in `BasePerformanceTest` to initialize the `setupzipper` dynamic variable with the correct context.
Something like this should do it:

    def defaultConfig: 


# Alternative B

axel22 replied:
It seems to me that the initial `prepareContext` call must:
a) do some side-effecting actions to prepare a testing environment
b) return a couple of values that describe the prepared testing environment

Initially, I was hoping we could replace the top-level context altogether before it gets used to create `Tree[Setup[T]]`, and that `BasePerformanceTest#defaultConfig` could be used for this purpose. Now I see that this is broken - `defaultConfig` is just ignored.

The way to fix it is to use `defaultConfig` in `BasePerformanceTest` to initialize the `setupzipper` dynamic variable with the correct context.
Something like this should do it:

    def defaultConfig = Context.empty // this is user-overridable
    final def initialConfig = measurer.prepareContext(currentContext ++ Context.empty)
    setupzipper.value = initialConfig

This should ensure that `Tree` objects have the correct contexts while the test tree is being constructed.
Eventually, it will bring the measurer-specific parameters to each and every test snippet.

The problem here is, of course, that contexts in the end contain things like the lookup table, which should not be persisted. I think the solution is to introduce transient keys, which we never persist. We seem to be hitting a fair amount of such keys anyway, and it's awkward to tweak the API in an effort to avoid them being serialized.

kjanosz replied:
Just for clarification, you've meant something like:
```
def defaultConfig = Context.empty // this is user-overridable
final def initialConfig = measurer.prepareContext(currentContext ++ defaultConfig) // you used Context.empty earlier
setupzipper.value = initialConfig
```
?

axel22 replied:
Yes! Sorry for the typo!

kjanosz replied:
One thing has come to my mind, also the `classpath` key will be changed - it will contain the instrumented jar. 
So the question is, should we clean the classpath (so the `cleanContext` method  should stay) or we don't care about this additional classpath entry?

kjanosz replied:
Also, about skipping some keys - it should be bypassed in all persistors? Or only in json based?

kjanosz replied:
Btw, since I should also adjust context in all setups is it OK to just call the `map` method on tree and substitute each setup context with result of the `initialConfig`?

axel22 replied:
Json based can easily bypass it.
Serialization-based persistor could have problems with it however. We
could address those in a separate PR.

On 7/9/2015 1:51 AM, Krzysztof Janosz wrote:
>>    def run[T](setuptree: Tree[Setup[T]], reporter: Reporter, persistor: Persistor): Tree[CurveData] = {
>> +    val context = measurer.prepareContext(setuptree.context)
> 
> Also, about skipping some keys - it should be bypassed in all persistors? Or only in json based? 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/scalameter/scalameter/pull/135/files#r34211586
> 

-- 
Aleksandar Prokopec

axel22 replied:
The classpath entry is just one additional key.
I'd say that it is not very dangerous. Let's leave it in for now.

On 7/9/2015 1:42 AM, Krzysztof Janosz wrote:
>>    def run[T](setuptree: Tree[Setup[T]], reporter: Reporter, persistor: Persistor): Tree[CurveData] = {
>> +    val context = measurer.prepareContext(setuptree.context)
> 
> One thing has come to my mind, also the `classpath` key will be changed - it will contain the instrumented jar. 
> So the question is, should we clean the classpath (so the `cleanContext` method  should stay) or we don't care about this additional classpath entry?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/scalameter/scalameter/pull/135/files#r34211064
> 

-- 
Aleksandar Prokopec

axel22 replied:
Careful - there is a Context on the setup class, but also a context on
the Tree itself. Mapping would only modify the context in Setup.

On 7/9/2015 2:32 AM, Krzysztof Janosz wrote:
>>    def run[T](setuptree: Tree[Setup[T]], reporter: Reporter, persistor: Persistor): Tree[CurveData] = {
>> +    val context = measurer.prepareContext(setuptree.context)
> 
> Btw, since I should also adjust context in all setups is it OK to just call the `map` method on tree and substitute each setup context with result of the `initialConfig`?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/scalameter/scalameter/pull/135/files#r34213878
> 

-- 
Aleksandar Prokopec

axel22 commented on src/main/scala/org/scalameter/Executor.scala:
```diff
@@ -1,18 +1,21 @@
     for (setup <- setuptree) yield {
       val exec = Option(setup.customExecutor).getOrElse(this)
-      val cd = exec.runSetup(setup)
+      val _cd = exec.runSetup(setup.copy(context = setup.context ++ context))
```
With the above changes, here you won't need to modify the context, nor clean it.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/Measurer.scala:
```diff
@@ -407,6 +418,47 @@ object Measurer {
+   *  @param primitives primitive types whose autoboxing will be counted.
+   *                    Note that leaving it empty means that autoboxing of all primitive types will be counted.
+   */
+  class AutoBoxingCounting(primitives: Class[_ >: Primitive]*) extends InvocationCounting {
```
Maybe just `BoxingCount`.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/Measurer.scala:
```diff
@@ -407,6 +418,47 @@ object Measurer {
+          if (primitives.nonEmpty)
+            new Regex(primitives.map(p => s"(${primitiveToBoxedMethod(p)})").mkString("^", "|", "$"))
+          else
+            "^boxTo\\w+$".r
```
This is actually surprising and somewhat unexpected.
I'd suggest to instead have a factory method that creates the `BoxingCount` for all the primitive types.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/Measurer.scala:
```diff
@@ -407,6 +418,47 @@ object Measurer {
+  }
+
+  /** Counts invocations of arbitrary method(s) specified by [[org.scalameter.execution.invocation.InvocationCountMatcher]]. */
+  class MethodInvocationCounting(val matcher: InvocationCountMatcher) extends InvocationCounting {
```
Maybe just `MethodInvocationCount`.

axel22 commented:
Regarding repackaging of the ASM, if you think it might cause API conflicts, we could move the new measurer implementations into a separate, optional ScalaMeter module. But I don't think we need to worry about this yet.

kjanosz commented:
@axel22: Looks like something is broken now. 
It's strange that this occurs only when running `sbt test` but not when running tests using `sbt testOnly` (I've initially thought that everything is OK because I was running locally tests singularly, not all at once.)
I will look at it and fix it tomorrow.

axel22 commented on src/main/scala/org/scalameter/PerformanceTest.scala:
```diff
@@ -38,11 +37,15 @@ object PerformanceTest {
       val datestart: Option[Date] = Some(new Date)
-      val rawsetuptree = BasePerformanceTest.setupzipper.value.result
+      val initialContext = executor.measurer.prepareContext(currentContext ++ defaultConfig)
+      val rawsetuptree = BasePerformanceTest.setupzipper.value.setContext(initialContext).result.map { setup =>
```
This is not doing the right thing - the `setupzipper` initialization process is a bit tricky, and is relying heavily on dynamic variables to achieve the nice dsl.

Take a look at the `dsl.scala` file. The parts of the tree in the `setupzipper` are initialized exactly when the various `performance of` and `using` blocks are called. This happens during the construction of the `BasePerformanceTest` subclass object, which is _before_ `executeTests` gets called.
When `executeTests` gets called, the setup zipper is already constructed, and has all the contexts and `Setup` values populated.

You must ensure that the initial value of the `setupzipper` is set correctly in the `BasePerformanceTest` constructor before the (user-defined) subclass constructors get called. By correct, I mean - the setup zipper must contain only empty the root tree value (`Tree.Zipper.root`) and the context of this root must be set to:

    prepareContext(currentContext ++ defaultConfig)

My advice is also that you change the `Tree.Zipper.root` factory method to take the `initialContext` argument, instead of relying on a dynamic variable.

kjanosz commented:
@axel22: Fixed.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/Measurer.scala:
```diff
@@ -12,6 +16,21 @@ trait Measurer extends Serializable {
   def units: String
+
+  /** Prepares a context that the setup context should be concatenated to.
+   *
```
Could you write:

    Modifies the initial test context.
    
    This method is invoked before the `PerformanceTest` object's constructor is 
    invoked.
    The key-value pairs that the `Measurer` adds to the Context in this method are 
    visible to all the test snippets within the `PerformanceTest` class.

    Most measurers do not need to add any specific keys, so the default 
    implementation just returns the `context`.

axel22 commented on src/main/scala/org/scalameter/PerformanceTest.scala:
```diff
@@ -43,6 +40,7 @@ object PerformanceTest {
       val rawsetuptree = BasePerformanceTest.setupzipper.value.result
       val setuptree = rawsetuptree.filter(setupFilter)
       val resulttree = executor.run(setuptree.asInstanceOf[Tree[Setup[SameType]]], reporter, persistor)
+      measurer.afterExecution(setuptree.context)
```
Technically, this would now mean that `executeTests` can only be invoked once.

You should try to invoke `bench:test` in SBT with the `fork in Bench`/`Test` set to `false`, and see what happens (my gut feeling is that the same test object is reused).

axel22 replied:
If this is true, then we should not call `afterExecution` here.

kjanosz commented:
@axel22: I will look at your comments and suggestions in about 2-3h and will try to fix this PR asap.

axel22 commented on scalameter-core/src/main/scala/org/scalameter/Measurer.scala:
```diff
@@ -407,6 +426,49 @@ object Measurer {
+  }
+
+  object BoxingCount {
+    def all() = new BoxingCount(classOf[Boolean], classOf[Char], classOf[Byte],
```
Could you add a ScalaDoc comment?

axel22 commented:
Sounds good, no rush!

axel22 commented on scalameter-core/src/main/scala/org/scalameter/PicklerBasedKey.scala:
```diff
@@ -5,20 +5,23 @@ import org.scalameter.picklers.Pickler
+  /** Pickler used to deserialize value to which the key refers. */
   def pickler: Pickler[T]
 
+  /** Indicates if a key should be serialized. */
```
Indicates if a key should be skipped during serialization - if `true`,
    the key will not be serialized.

axel22 commented on src/main/scala/org/scalameter/execution/JvmRunner.scala:
```diff
@@ -32,7 +32,10 @@ final class JvmRunner {
   private def runJvm(ctx: Context) {
-    val classpath = ctx.goe(Key.classpath, utils.ClassPath.default)
+    val classpath = {
+      val cl = ctx.goe(Key.classpath, utils.ClassPath.default)
```
Could we divide the keys into `Key.nonModifiedClasspath` and `Key.classpath` - then `nonModifiedClasspath` would not contain `instrumentedJarPath`, and `classpath` would contain any additions by measurers, executors, etc.

This could reduce the chances of somebody forgetting to add `instrumentedJarPath`. Now we apparently only have `JvmRunner`, but we might have a `ClassLoaderRunner` in the future (which creates a sub-classloader).

We can do this in a separate PR, though.

axel22 commented on src/main/scala/org/scalameter/execution/LocalExecutor.scala:
```diff
@@ -27,6 +27,7 @@ import utils.Tree
  *  the young generation.
  */
 class LocalExecutor(val warmer: Warmer, val aggregator: Aggregator, val measurer: Measurer) extends Executor {
+  require(!measurer.isInstanceOf[InvocationCount], "InvocationCount measurers should be run using SeparateJvmsExecutor")
```
This is a bit ad-hoc. I would add a method `usesInstrumentedClasspath` to `Measurer`, which defaults to `false`.

axel22 commented on src/test/scala/org/scalameter/MethodInvocationCountTest.scala:
```diff
@@ -0,0 +1,7 @@
+
+class MethodInvocationCountTest extends InvocationCountMeasurerTest {
+  test("MethodInvocationCountTest.allocations should be deterministic") {
+    checkInvocationCountMeasurerTest(new org.scalameter.examples.MethodInvocationCountTest1)
```
Could you invoke this twice on the same object during the test (particularly if SBT proves problematic about re-running the test, as I suggested in the earlier comment).

kjanosz replied:
Ach, my bad - you're right. To fix it, overriding `finalize` and doing cleanup there (with additional `jar.deleteOnExit` right after creation) would be sufficient?

kjanosz replied:
This seems to work OK (check only calling manually `finalize` after a few `executeTests` because JVM ignored `System.gc`)
```
override def finalize(): Unit = {
  measurer.afterExecution(currentContext)
  super.finalize(BasePerformanceTest.setupzipper.value.result.context)
}
```

axel22 replied:
I wouldn't use `finalize` here - it is too brittle. What we really want is:
1. reserve a temporary file before the test ctor is called, so that context can get a new path - done by the `Measurer#prepareContext` in `BasePerformanceTest`.
2. use the temporary file to instrument the classpath just before tests are executed - done by `Measurer#beforeTests` in `executeTests`.
3. delete the temporary jar file - done by `Measurer#afterTests`.

This way, the same temporary file path is reused multiple times, but created and deleted right before and after the tests start/complete.
I think this is the right design.

kjanosz replied:
I think that the problem arises with the lookup table key. It must be in the context before invocation count `measure` method and it is created after writing instrumented classes. So either `Instrumentation.writeInstrumentedClasses` must be refactored to do some in-memory check (but I think that this could be inefficient for more computationally expensive matchers, like the on-way `Descendants` class matcher), or it should insert `methodInvocationLookupTable` into the context before calling measurer `measure` (so in the executor or in performance test (?))
Or do you see other alternatives?

axel22 replied:
I would do that - insert all the necessary keys early, in `prepareContext`, and then perform side-effects immediately before the tests start in `beforeContext.

kjanosz replied:
Btw, I've just noticed that `Initialization` and `JavaPerformanceTest` share the same `executeTests` body - could it be a good idea to just move it to one base class and declare `executeTests` as final?

axel22 replied:
Indeed - as far as I remember, there is no specific reason for this, and it is just an oversight. Nice catch - please feel free to move it into `BasePerformanceTest` and make it `final`.

kjanosz commented:
@axel22: Changed.
I wasn't sure where to place `afterExecution` - now it's placed in `finalize` method of a `BasePerformanceTest`.
I've also added a shutdown hook (I've removed old line `jar.deleteOnExit()` from `InvocationCount.prepareContext` since it works in a similar way).

axel22 commented:
I left comment above. Sorry that we're going a bit back-and-forth, but I think what we have after the last comment above should be the right design.

kjanosz commented:
@axel22: Now everything should be OK (I've fixed failed tests and done refactor with `executeTests`)

axel22 commented:
Awesome work - LGTM!
I will wait until the tests pass, and then I'm merging this in!
